ddp-workspace/projects/ddp-circadia/src/app/services/sleep-log.service.ts

```diff
@@ -0,0 +1,193 @@
+  diaryUrl$ = new BehaviorSubject<string | null>(null);
+  private userEmail: string | null = null;
+  private initialized$ = new BehaviorSubject<boolean>(false);
+
+  constructor(
+    private http: HttpClient,
+    private datePipe: DatePipe,
+    private sessionService: SessionMementoService,
+    private auth0Service: Auth0AdapterService,
+    private loggingService: LoggingService,
+    private announcementsService: AnnouncementsServiceAgent,
+    @Inject('ddp.config') private config: ConfigurationService,
+  ) {
+    this.sessionService.sessionObservable
+      .pipe(
+        filter(session => !!session && !!session.accessToken),
+        tap(({ accessToken }) =>
+          this.auth0Service.webAuth.client.userInfo(
+            accessToken,
+            (err, user) => {
+              if (err) {
+                this.loggingService.logError(
+                  SleepLogService.LOG_SOURCE,
+                  'Error occured when tried to fetch user info',
+                  err,
+                );
+
+                return;
+              }
+
+              this.userEmail = user.name;
+              this.initialized$.next(true);
+            },
+          ),
+        ),
+        take(1),
+      )
+      .subscribe();
+  }
+
+  extractSleepLogAnnouncements(
+    announcements: AnnouncementMessage[],
+  ): SleepLogCode[] {
+    const sleepLogAnnouncements: SleepLogCode[] = [];
+
+    announcements.forEach(announcement => {
+      if (SleepLogService.SLEEP_LOG_CODES.includes(announcement.message)) {
+        sleepLogAnnouncements.push(announcement.message as SleepLogCode);
+      }
+    });
+
+    return sleepLogAnnouncements;
+  }
+
+  runSleepLogActions(
+    codes: SleepLogCode[],
+    hasSleepLogActivity: boolean,
+  ): Observable<CreateUserResponse | UserInfoResponse | null> {
+    let actions: (SleepLogCode | UtilitySleepLogCode)[] = [];
+
+    if (!codes.length) {
+      if (hasSleepLogActivity) {
+        actions = [UtilitySleepLogCode.GetInfo];
+      }
+    } else if (!codes.includes(SleepLogCode.CreateUser)) {
+      actions = [...actions, UtilitySleepLogCode.GetInfo];
+    } else {
+      actions = codes;
+    }
+
+    return this.initialized$.pipe(
+      filter(Boolean),
+      take(1),
+      concatMap(() => from(actions)),
+      concatMap(action => this.createSleepLogAction(action)),
+    );
+  }
+
+  private createSleepLogAction(
+    code: SleepLogCode | UtilitySleepLogCode,
+  ): Observable<any> {
+    switch (code) {
+      case SleepLogCode.CreateUser:
+        return this.createUser();
+      case UtilitySleepLogCode.GetInfo:
+        return this.getUserInfo();
+      default:
+        return of(null);
+    }
+  }
+
+  private createUser(): Observable<CreateUserResponse | null> {
+    const payload: CreateUserPayload = {
+      ...this.getCommonProxyPayload(HttpMethod.Post),
+      start: this.datePipe.transform(Date.now(), SleepLogService.DATE_FORMAT),
+      end: this.datePipe.transform(
+        Date.now() +
```

Comment:
You sure you want to get time from user's browser instead of from server? Worrying about being dependent on local computer time being accurate/and or consistent across all users.

Reply:
I guess we don't have an alternative unless an API changes what it accepts in request's payload.

Reply:
Do we control the back-end service? Or is this an external service provided by our collaborators? To follow up on the date business, if we implement the back-end, we should change the back-end to get its own date.

Reply:
I removed this.